### PR TITLE
feat(eslint-plugin): [prefer-optional-chain] support logical with empty object

### LINKED ISSUE #4395: [prefer-optional-chain] Suggest `(foo || {}).bar` as a valid match for the rule
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-optional-chain": "error",
  }
}
```

```TS
const foo = (bar || {}).baz;
// or
const foo = (bar|| {})[baz];
// or
const foo = (((a || {}).b || {}).c || {}).d;
```

**Expected Result**
The rule should suggest to use optional chaining

I'm working on a codebase that uses this pattern a lot (about ~60% of the cases I refactor to optional chaining arise from this pattern, the others come from `&&`s) and I'm only getting prompted for the `&&` cases.

I expect to get the same warnings/errors/suggestions the `&&` case have, not including function calls

**Actual Result**
No error or warning

**Additional Info**
Willing to add a PR if needed
Tested this already against existing code in this repo

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.9.0` |
| `@typescript-eslint/parser`        | `5.9.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.1.0` |
| `node`                             | `14.17.5` |

### PR DESCRIPTION
**Pull request by omril1**
Adds empty object pattern (`(foo || {}).bar` or `(foo ?? {}).bar`) as a match for `prefer-optional-chain` rule

## PR Checklist

-   [x] Addresses an existing issue: fixes #4395
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Adds support for suggesting optional chaining refactor when an empty object is used to mimic optional chaining in addition to the existing `&&` pattern (`(foo || {}).bar` and `(foo ?? {}).bar` are roughly equivalent to `foo && foo.bar` or `foo?.bar`)

## Notes

- The feature does not uses type information
- The feature does not handle function calls in this way as `(foo || {}).bar()` can cause a runtime error (unlike the match for `&&`).

### CODE REVIEW

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (B2o5T) [inline packages/eslint-plugin/src/rules/prefer-optional-chain.ts:73]: sourceCode already available in the upper scope at line 49

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (B2o5T) [inline packages/eslint-plugin/src/rules/prefer-optional-chain.ts:79]: Same remark

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts:73]: Thanks, didn't notice

File: packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts:151]: I think this is a smart thing to add actually 😄. Checking for unnecessary optional chains is part of `no-unnecessary-condition`'s responsibility. 

```suggestion
    '(foo || {})?.bar;',
```

File: packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts:543]: Nit: I think this is implied by looking at the test case? IME these comments can fall out of sync pretty quickly.

```suggestion
```

File: packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts:509]: These test cases are looking good so far! Can you add some for:

* Combinations of `?.` and `{}`, like the `'(foo1?.foo2 || {}).foo3;'` output below
* `||=` (just to make sure we don't accidentally catch that)
* `'(foo || {}).bar` itself
* A few more types of expressions to the left of the `||`: maybe, a function call, a ternary, one or two others? (making sure edge cases don't sneak in)
* Inside an `if` statement

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/prefer-optional-chain.ts:51]: > The feature does not handle function calls in this way as `(foo || {}).bar()` can cause a runtime error.

Hmm, interesting point. My hunch would be that we probably want to handle function calls for cases like this:

```ts
(someContainer || {}).hasOwnProperty("someKey")
```

...but I wouldn't consider it a blocker / must-have for this PR. Up to you!

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Looks great so far, thanks!

Just requesting changes on a bit more testing please ❤️

File: packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts
Comment (omril1) [reply packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts:509]: I think it does have some complications with a ternary to the left of `||`
* if the ternary is wrapped with parenthesis the output is wrong as it loses the wrapping parenthesis
(`((foo1 ? foo2 : foo3) || {}).foo4` -> `foo1 ? foo2 : foo3.foo4` which is definitely wrong when `foo2` is falsy), good catch on that.
* if the ternary is not wrapped (`(foo1 ? foo2 : foo3 || {}).foo4`) there are no reported issues which I think is the most correct for most cases (if we report this it definitely has higher false positive rates)

Comment (omril1) [reply packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts:509]: Do you think `(foo ?? {}).bar` should also match? It's likely more accurate than `||` in that it only picks `null` or `undefined`, but it's a more modern, same as optional chaining

File: packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts
Comment (omril1) [reply packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts]: It was not obvious to me as I was not familiar with how the fixer works when there are 2 fixes, does it only picks the first one?
I removed the comment for now.

File: packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts:521]: Ha good callout. Yes I think we should handle that -- it's technically closer to `?.`!

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts]: @JoshuaKGoldberg Care to re-review this? I don't think I'm going to add object built-ins in the PR

File: packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/prefer-optional-chain.test.ts]: I forget what the default behavior in tests is but it'd be the same across all rule testers. We never customize it.

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts]: You got it, no problem!

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts]: You got it, no problem!

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/prefer-optional-chain.ts:51]: This _seems_ right to me, but the rule elsewhere has a bunch of checks on the kinds of binary expressions in play so I'd want to page in @bradzacher for a double check before I merge this.

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/prefer-optional-chain.ts:56]: Just nitpicking: I can see why you made the two bigger variables below to make the code more clear, but at least for these three the single character of name savings / object member lookup isn't worth the extra code bloat.

Not a blocker IMO (someone should eventually write a lint rule if it's important), just a thought. :)

**Review by JoshuaKGoldberg — APPROVED** [review]
Looks great to me, thanks so much for the thorough implementation & discussion! 🎉 

@bradzacher is all over the history for this already-complex file so I'll just want to get a second opinion before merging.

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts:56]: It started out with `parentNode` for reuse and typescript's type narrowing inside the fixer (typescript doesn't narrow property's type inside other functions' scopes if the type guard is above the function).

The left & right nodes are just a stylistic choice to match it.

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts:51]: @JoshuaKGoldberg @bradzacher Any updates?

Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts:51]: I also think [DeMorgan’s Laws](https://ntgard.medium.com/https-medium-com-ntgard-boolean-logic-in-javascript-part-2-ec0b13169b2#:~:text=!(A%20%7C%7C%20B)%20%3D%3D%3D%20!A%20%26%26%20!B) can also be applied to the rule to cover more cases.
`!foo || !foo.bar` (no fix suggest) is equivalent to `!(foo && foo.bar)` (suggests `!(foo?.bar)`)
Should I open an issue for it?

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts:51]: (nitpick on styling)
```suggestion
      'LogicalExpression[operator="||"], LogicalExpression[operator="??"]'(
```

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/prefer-optional-chain.ts:63]: I don't think it's worth having this check.
This is a deep rabbit-hole to go down to statically analyse all of the operators to determine if they could possibly return a nullish value.
Only handling a subset of them feels wrong, because it means the rule won't check some cases and will check others - to a user it just feels janky to be inconsistent.

There's also things like `(typeof x as string) || {}` which would be ignored because your code doesn't "look past" the assertion.
Another example of the rabbit-hole - your check will false-negative on `void x` - which is a `UnaryExpression` that always returns `undefined`.

I would just go with the same logic of the rest of the rule - which is to always assume the left side could be nullable.

---

If, however, you wanted to implement something more thorough - there is an example code like this which could be reworked and built on top of
https://github.com/eslint/eslint/blob/97a4f6385cbcd8ea251de17dc2f1683132c2d0d9/lib/rules/no-constant-binary-expression.js#L23-L84

The basic idea would be to do this:
1) if the thing is an `Identifier`, use scope analysis to find the definition
   a) if the definition is variable with more than one assignment, assume it could be nullable.
   b) else analyse the variable's singular written value
2) if the value is statically non-null, then skip it
3) if the value is potentially nullish, continue evaluating.

**Review by bradzacher — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts:63]: I had good intentions to not show the lint error on things that are clearly messed up, but the rule cannot fix or warn correctly, like values that cannot be objects (accessing properties on a boolean or number just goes to the prototype, and most of them are methods that could better be accessed directly)

I don't think I'm familiar enough with the API of how to perform scope analysis and I assume it's not a trivial change.
If you think it's enough to drop the check I'll go with that approach.

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts:63]: Definitely agree with the good intentions!
I think for now we can drop the check for this first commit, and then we can come back and assess to see if we can holistically improve this rule with as many cases as we can.

Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts]: Hi, @bradzacher & @JoshuaKGoldberg Is there anything else I need to change?
I think we can resolve the requested changes.

**Comment by omril1** [comment]
@bradzacher & @JoshuaKGoldberg any update about this PR?

**Comment by bradzacher** [comment]
it's on the review queue for when the volunteer maintainers have time to review PRs!

**Review by bradzacher — CHANGES_REQUESTED** [review]
looking good! Just one comment

File: packages/eslint-plugin/src/rules/prefer-optional-chain.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/prefer-optional-chain.ts]: OK, I did that with a slight alteration because it didn't work (even after fixing the variable names from the snippet)

**Review by bradzacher — APPROVED** [review]
LGTM - thanks for thsi!

**Comment by avaly** [comment]
This change seems to be a breaking change because this rule now requires type information.

**Comment by bradzacher** [comment]
As per our contributing guide. Please don't comment on closed PRs.
Instead raise an issue and fill out the template.

**Pull request merged**
